lib/internal/streams/web.js

```diff
@@ -0,0 +1,4632 @@
+
+    if (viewByteLength === 0 || viewBufferByteLength === 0) {
+      throw new ERR_INVALID_STATE.TypeError(
+        'View ArrayBuffer is zero-length or detached');
+    }
+
+    if (controller === undefined) {
+      throw new ERR_INVALID_STATE.TypeError(
+        'This BYOB request has been invalidated');
+    }
+
+    readableByteStreamControllerRespond(controller, bytesWritten);
+  }
+
+  /**
+   * @param {ArrayBufferView} view
+   */
+  respondWithNewView(view) {
+    if (!isReadableStreamBYOBRequest(this))
+      throw new ERR_INVALID_THIS('ReadableStreamBYOBRequest');
+    const {
+      controller,
+    } = this[kState];
+
+    if (controller === undefined) {
+      throw new ERR_INVALID_STATE.TypeError(
+        'This BYOB request has been invalidated');
+    }
+
+    readableByteStreamControllerRespondWithNewView(controller, view);
+  }
+
+  [kInspect](depth, options) {
+    return customInspect(depth, options, 'ReadableStreamBYOBRequest', {
+      view: this.view,
+      controller: this[kState].controller,
+    });
+  }
+}
+
+class InternalReadableStreamBYOBRequest {
+  constructor(controller, view) {
+    this[kType] = 'ReadableStreamBYOBRequest';
+    this[kState] = {
+      controller,
+      view,
+    };
+  }
+}
+
+class DefaultReadRequest {
+  constructor() {
+    this[kState] = createDeferredPromise();
+  }
+
+  [kChunk](value) {
+    this[kState].resolve?.({ value, done: false });
+  }
+
+  [kClose]() {
+    this[kState].resolve?.({ value: undefined, done: true });
+  }
+
+  [kError](error) {
+    this[kState].reject?.(error);
+  }
+
+  get promise() { return this[kState].promise; }
+}
+
+class ReadIntoRequest {
+  constructor() {
+    this[kState] = createDeferredPromise();
+  }
+
+  [kChunk](value) {
+    this[kState].resolve?.({ value, done: false });
+  }
+
+  [kClose](value) {
+    this[kState].resolve?.({ value, done: true });
+  }
+
+  [kError](error) {
+    this[kState].reject?.(error);
+  }
+
+  get promise() { return this[kState].promise; }
+}
+
+class ReadableStreamDefaultReader {
+  [kType] = 'ReadableStreamDefaultReader';
+
+  /**
+   * @param {ReadableStream} stream
+   */
+  constructor(stream) {
+    if (!isReadableStream(stream))
+      throw new ERR_INVALID_ARG_TYPE('stream', 'ReadableStream', stream);
+    this[kState] = {
+      readRequests: [],
+      stream: undefined,
+      close: {
+        promise: undefined,
+        resolve: undefined,
+        reject: undefined,
+      },
+    };
+    setupReadableStreamDefaultReader(this, stream);
+  }
+
+  /**
+   * @returns {Promise<{
+   *   value : any,
+   *   done : boolean
+   * }>}
+   */
+  read() {
+    if (!isReadableStreamDefaultReader(this))
+      throw new ERR_INVALID_THIS('ReadableStreamDefaultReader');
+    if (this[kState].stream === undefined) {
+      return PromiseReject(
+        new ERR_INVALID_STATE.TypeError(
+          'The reader is not attached to a stream'));
+    }
```

Comment:
```suggestion
  async read() {
    if (!isReadableStreamDefaultReader(this))
      throw new ERR_INVALID_THIS('ReadableStreamDefaultReader');
    if (this[kState].stream === undefined) {
      throw new ERR_INVALID_STATE.TypeError(
        'The reader is not attached to a stream');
    }
```

currently, this function sometimes throws a sync exception and sometimes returns a Promise, which is both zalgo and something the JS spec tries *very hard* not to do. The suggested change means that it only ever returns a promise and never synchronously throws.

Reply:
It's also non-compliant to throw a sync exception here. 😉 Since `read()` returns a promise type, ["create an operation function"](https://heycam.github.io/webidl/#dfn-create-operation-function) requires that this type check from step 2.1.2.3:
> 3. If _esValue_ does not implement the interface _target_, throw a TypeError.

is caught and turned into a rejection:
> And then, if an exception _E_ was thrown:
> 1. If _op_ has a return type that is a promise type, then return ! Call(`%Promise.reject%`, `%Promise%`, «_E_»).

Same thing applies to `readable.cancel()`, `reader.cancel()`, `reader.closed`, `writable.abort()`, `writable.close()`, `writer.abort()`, `writer.close()`, `writer.write()` `writer.closed` and `writer.ready`. The IDL harness tests should cover all these cases.